evm/src/cpu/kernel/asm/curve/bls381/util.asm

```diff
@@ -120,4 +199,125 @@ global mul_fp381_2:
+    %pop3
+    JUMP
+
+global sub_fp381_6:
+    // stack:           inA, inB, out, jumpdest  { out: [ 0,  0,  0,  0,  0,  0 ] }
+    %sub_term_kernel                          
+    // stack:           inA, inB, out, jumpdest  { out: [C0,  0,  0,  0,  0,  0 ] }
+    %sub_term_kernel(2)                       
+    // stack:           inA, inB, out, jumpdest  { out: [C0, C1,  0,  0,  0,  0 ] }
+    %sub_term_kernel(4)                       
+    // stack:           inA, inB, out, jumpdest  { out: [C0, C1, C2,  0,  0,  0 ] }
+    %sub_term_kernel(6)                       
+    // stack:           inA, inB, out, jumpdest  { out: [C0, C1, C2, C3,  0,  0 ] }
+    %sub_term_kernel(8)                       
+    // stack:           inA, inB, out, jumpdest  { out: [C0, C1, C2, C3, C4,  0 ] }
+    %sub_term_kernel(10)                      
+    // stack:           inA, inB, out, jumpdest  { out: [C0, C1, C2, C3, C4, C5 ] }
+    %pop3
+    JUMP
+
+%macro add_term_kernel
+    // stack:           inA, inB, out, jumpdest
+    DUP2
+    // stack:     inB0, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:       B0, inA, inB, out, jumpdest
+    DUP3
+    // stack: inA0, B0, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:   A0, B0, inA, inB, out, jumpdest
+    %add_fp381
+    // stack:       C0, inA, inB, out, jumpdest
+    DUP5
+    // stack: out0, C0, inA, inB, out, jumpdest
+    %mstore_bls
+%endmacro
+
+%macro add_term_kernel(n)
+    // stack:           inA, inB, out, jumpdest
+    DUP2
+    %add_const($n)
+    // stack:     inBn, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:       Bn, inA, inB, out, jumpdest
+    DUP3
+    %add_const($n)
+    // stack: inAn, Bn, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:   An, Bn, inA, inB, out, jumpdest
+    %add_fp381
+    // stack:       Cn, inA, inB, out, jumpdest
+    DUP5
+    %add_const($n)
+    // stack: outn, Cn, inA, inB, out, jumpdest
+    %mstore_bls
+%endmacro
+
+%macro sub_term_kernel
+    // stack:           inA, inB, out, jumpdest
+    DUP2
+    // stack:     inB0, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:       B0, inA, inB, out, jumpdest
+    DUP3
+    // stack: inA0, B0, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:   A0, B0, inA, inB, out, jumpdest
+    %sub_fp381
+    // stack:       C0, inA, inB, out, jumpdest
+    DUP5
+    // stack: out0, C0, inA, inB, out, jumpdest
+    %mstore_bls
+%endmacro
+
+%macro sub_term_kernel(n)
+    // stack:           inA, inB, out, jumpdest
+    DUP2
+    %add_const($n)
+    // stack:     inBn, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:       Bn, inA, inB, out, jumpdest
+    DUP3
+    %add_const($n)
+    // stack: inAn, Bn, inA, inB, out, jumpdest
+    %mload_bls
+    // stack:   An, Bn, inA, inB, out, jumpdest
+    %sub_fp381
+    // stack:       Cn, inA, inB, out, jumpdest
+    DUP5
+    %add_const($n)
+    // stack: outn, Cn, inA, inB, out, jumpdest
+    %mstore_bls
+%endmacro
```

Comment:
I don't think you need to specialize the `n = 0` cases. I believe the assembler should be able to optimize that out (pls correct me if I'm wrong @dlubarov).